src/runner/program/mod.rs

```diff
@@ -0,0 +1,110 @@
+mod cache;
+
+mod shell;
+use shell::*;
+mod rust;
+use rust::*;
+mod python;
+use python::*;
+
+use std::borrow::Cow;
+use thiserror::Error;
+use super::CodeBlock;
+
+trait Program {
+    fn run(&self, code_block: &CodeBlock) -> Result<std::process::Output, std::io::Error>;
+}
+
+#[derive(Error, Debug)]
+pub enum RunError {
+    #[error("error while searching a program: {0}")]
+    Search(#[from] SearchError),
+    #[error("program not found for `{0}`")]
+    ProgramNotFound(String),
+    #[error("io error: {0}")]
+    Io(#[from] std::io::Error),
+}
+#[derive(Error, Debug)]
+pub enum SearchError {
+    #[error("env var {0} not found: {1}")]
+    EnvVarNotFound(Cow<'static, str>, std::env::VarError),
+    #[error("bad utf-8 encoding in path")]
+    BadUTF8,
+    #[error("no corresponding program found for `{0}`")]
+    NoCorrespondingProgram(String),
+    #[error("cache error: {0}")]
+    Cache(#[from] cache::CacheError)
+}
+
+enum SearchStatus {
+    Found(Box<dyn Program>),
+    NotFound,
+    Error(SearchError)
```

Comment:
isn't `NotFound` a kind of Error ?

Reply:
It can be. But Errors are differently managed where the search is queried. if found or not found, the function continue to run, but if it is an Error, it returns.
At first, `SearchStatus` was a `Result<Option<Box<dyn Program>, SearchError>`. As you can see, this is pretty much the same structure.